### PR TITLE
Nodejs week2

### PR DESCRIPTION
**Pull request by andrii1**

### CODE REVIEW

File: src/backend/api/meals.js [hunk 1]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
 const router = express.Router();
 const knex = require("../database");
 
+// /api/meals	GET	Returns all meals
+
 router.get("/", async (request, response) => {
   try {
     // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
+    const titles = await knex("meal").select("title");
```
Comment (orhantoy) [inline src/backend/api/meals.js]: This would only return the `title` column of the meals.

File: src/backend/api/meals.js [hunk 2]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
...
+
+router.post("/", async (request, response) => {
+  const newMeal = await knex("meal").insert({
+    title: request.body.title,
+    description: request.body.description,
+    location: request.body.location,
+    when: request.body.when,
+    max_reservations: request.body.max_reservations,
+    price: request.body.price,
+    created_date: request.body.created_date,
```
Comment (orhantoy) [inline src/backend/api/meals.js]: I don't think the user should be able to determine when the meal was created.

File: src/backend/api/reservations.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  response
+    .status(201)
+    .json({ message: "Created a new reservation", id: newReservation[0] });
+});
+
+// /api/reservations/:id	GET	Returns a reservation by id
+
+router.get("/:id", async (request, response) => {
+  try {
+    const idAvailable = await checkIfIdAvailable(request.params.id);
```
Comment (orhantoy) [inline src/backend/api/reservations.js:37]: In this route you're now performing 2 SQL queries but we technically only need 1.

File: src/backend/api/reservations.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  }
+});
+// /api/reservations/:id	PUT	Updates the reservation by id
+
+router.put("/:id", async (request, response) => {
+  try {
+    const idAvailable = await checkIfIdAvailable(request.params.id);
+
+    if (!idAvailable) {
+      response.status(400).json({ error: "id not available" });
```
Comment (orhantoy) [inline src/backend/api/reservations.js]: A 404 response makes more sense when we cannot find a meal with the given ID.

File: src/backend/api/reservations.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
+const express = require("express");
+const router = express.Router();
+const knex = require("../database");
+
+// /api/reservations	GET	Returns all reservations
+
+router.get("/", async (request, response) => {
+  try {
+    // knex syntax for selecting things. Look up the documentation for knex for further info
+    const titles = await knex("reservation");
```
Comment (orhantoy) [inline src/backend/api/reservations.js]: Calling this variable `titles` is a bit confusing. `reservations` would make more sense.

File: src/backend/api/reservations.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  }
+});
+
+// /api/reservations	POST	Adds a new reservation to the database
+
+router.post("/", async (request, response) => {
+  const newReservation = await knex("reservation").insert({
+    number_of_guests: request.body.number_of_guests,
+    meal_id: request.body.meal_id,
+    created_date: request.body.created_date,
```
Comment (orhantoy) [inline src/backend/api/reservations.js]: I would not allow the user to set this field.

File: src/backend/api/reservations.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    if (!idAvailable) {
+      response.status(400).json({ error: "id not available" });
+      return;
+    }
+    const updatedReservation = await knex("reservation")
+      .where({ "reservation.id": request.params.id })
+      .update({
+        number_of_guests: request.body.number_of_guests,
+        meal_id: request.body.meal_id,
+        created_date: request.body.created_date,
```
Comment (orhantoy) [inline src/backend/api/reservations.js]: I would not allow the user to set this field.

File: src/backend/api/reservations.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  }
+});
+// /api/reservations/:id	DELETE	Deletes the reservation by id
+
+router.delete("/:id", async (request, response) => {
+  try {
+    const idAvailable = await checkIfIdAvailable(request.params.id);
+
+    if (!idAvailable) {
+      response.status(400).json({ error: "id not available" });
```
Comment (orhantoy) [inline src/backend/api/reservations.js]: Similar to my other comment, a 404 response makes more sense here.

File: src/backend/app.js [hunk 9]
```diff
@@ -20,11 +31,52 @@ app.use(express.json());
...
+router.use("/reservations", reservationsRouter);
+
+//testing knex connection
+router.get("/info", async (req, res) => {
+  const dbResult = await knex.raw("SELECT VERSION()");
+  const row = dbResult[0][0];
+  res.json({ nodeVersion: process.version, mysqlVersion: row["VERSION()"] });
+});
+
+router.get("/future-meals", async (req, res) => {
```
Comment (orhantoy) [inline src/backend/app.js:43]: I'm not reviewing these routes from week 1.

File: src/backend/api/meals.js [hunk 10]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
...
+    const updatedMeal = await knex("meal")
+      .where({ "meal.id": request.params.id })
+      .update({
+        title: request.body.title,
+        description: request.body.description,
+        location: request.body.location,
+        when: request.body.when,
+        max_reservations: request.body.max_reservations,
+        price: request.body.price,
+        created_date: request.body.created_date,
```
Comment (orhantoy) [inline src/backend/api/meals.js]: I would not allow the user to set this field.

File: src/backend/api/meals.js [hunk 11]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
...
+  }
+});
+// /api/meals/:id DELETE	Deletes the meal by id
+
+router.delete("/:id", async (request, response) => {
+  try {
+    const idAvailable = await checkIfIdAvailable(request.params.id);
+
+    if (!idAvailable) {
+      response.status(400).json({ error: "id not available" });
```
Comment (orhantoy) [inline src/backend/api/meals.js]: Similar to my other comment, a 404 response makes more sense here.

File: src/backend/api/meals.js [hunk 12]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
...
+    created_date: request.body.created_date,
+  });
+
+  response.status(201).json({ message: "Created a new meal", id: newMeal[0] });
+});
+
+// /api/meals/:id	GET	Returns the meal by id
+router.get("/:id", async (request, response) => {
+  try {
+    const idAvailable = await checkIfIdAvailable(request.params.id);
```
Comment (orhantoy) [inline src/backend/api/meals.js:35]: We don't need to perform 2 SQL queries in this route.

File: src/backend/api/meals.js [hunk 13]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
...
+    throw error;
+  }
+});
+// /api/meals/:id	PUT	Updates the meal by id
+router.put("/:id", async (request, response) => {
+  try {
+    const idAvailable = await checkIfIdAvailable(request.params.id);
+
+    if (!idAvailable) {
+      response.status(400).json({ error: "id not available" });
```
Comment (orhantoy) [inline src/backend/api/meals.js]: Similar to my other comment, a 404 response makes more sense here.

File: src/backend/api/reservations.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        number_of_guests: request.body.number_of_guests,
+        meal_id: request.body.meal_id,
+        created_date: request.body.created_date,
+        contact_phonenumber: request.body.contact_phonenumber,
+        contact_name: request.body.contact_name,
+        contact_email: request.body.contact_email,
+      });
+    response.json({
+      message: "reservation updated",
+      id: updatedReservation[0],
```
Comment (orhantoy) [inline src/backend/api/reservations.js:72]: I don't think this returns the ID of the reservation but instead the number of rows updated, which is usually just 1.

**Review by orhantoy — COMMENTED** [review]
Looks great, Andrii! 💪 

The main thing to improve is some of the responses that should've been 404 and not 400.

File: src/backend/api/meals.js (hunk 1 above)
Comment (27ManishaPatel) [inline src/backend/api/meals.js]: Use SELECT '*' here

File: src/backend/api/meals.js [hunk 15]
```diff
@@ -2,14 +2,99 @@ const express = require("express");
...
+      return;
+    }
+    const meal = await knex("meal")
+      .where({ "meal.id": request.params.id })
+      .limit(1);
```
Comment (27ManishaPatel) [inline src/backend/api/meals.js:43]: id is unique then Is there any specific purpose to use LIMIT?

File: src/backend/api/reservations.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+  }
+});
+
+async function checkIfIdAvailable(id) {
+  let users = [];
+  users = await knex("reservation").where({ "reservation.id": id }).limit(1);
+  return users[0];
+}
+
+module.exports = router;
```
Comment (27ManishaPatel) [inline src/backend/api/reservations.js:106]: good work with reservations !

**Review by 27ManishaPatel — APPROVED** [review]
Well done !

**Comment by andrii1** [comment]
> 

thanks! for some reason I put 400 error, logic was "Invalid inputs by the user (server responds with a 400 response code);" https://www.smashingmagazine.com/2020/08/error-handling-nodejs-error-classes/

File: src/backend/api/meals.js (hunk 1 above)
Comment (andrii1) [reply src/backend/api/meals.js]: fixed

File: src/backend/api/meals.js (hunk 2 above)
Comment (andrii1) [reply src/backend/api/meals.js]: changed this in table ` ALTER TABLE meal CHANGE created_date created_date date NOT NULL DEFAULT (CURRENT_DATE)`

File: src/backend/api/reservations.js (hunk 6 above)
Comment (andrii1) [reply src/backend/api/reservations.js]: set in reservations table `DEFAULT (CURRENT_DATE)` and removed from js code

File: src/backend/api/reservations.js (hunk 7 above)
Comment (andrii1) [reply src/backend/api/reservations.js]: fixed

File: src/backend/api/reservations.js (hunk 3 above)
Comment (andrii1) [reply src/backend/api/reservations.js:37]: hm, what is the better way to do it? It is similar to the classwork we did

File: src/backend/api/meals.js (hunk 1 above)
Comment (andrii1) [reply src/backend/api/meals.js]: added this const meals = await knex("meal");

File: src/backend/api/meals.js (hunk 15 above)
Comment (andrii1) [reply src/backend/api/meals.js:43]: good notice, hm don't know

**Pull request merged**